File: src/0046.permutations.46/solution.ts
function permute (nums: number[]): number[][] {
    const result:number[][] = [];
    backTracking(
        nums,
        new Array(nums.length).fill(false),
        [] as number[],
        result
    );
    return result;
}

function backTracking (nums:number[], used:boolean[], sequence:number[], result:number[][]) {
    if (sequence.length === nums.length) {
        result.push(sequence.slice());
        return;
    }
    for (let i = 0; i < nums.length; i++) {
        if (used[i]) {
            continue;
        }
        sequence.push(nums[i]);
        used[i] = true;
        backTracking(nums, used, sequence, result);
        used[i] = false;
        sequence.pop();
    }
}

File: src/0046.permutations.46/solution.java
class Solution {
    public List<List<Integer>> permute(int[] nums) {
        boolean[] used = new boolean[nums.length];
        List<List<Integer>> result = new ArrayList<>();
        backTracking(nums,used,new ArrayList<>(),result);
        return result;
    }

    private void backTracking(int[] nums,boolean[] used,List<Integer> accumulate,List<List<Integer>> result){

        if(accumulate.size() == nums.length){
            result.add(new ArrayList<>(accumulate));
            return;
        }

        for(int i=0;i<nums.length;i++){
            if(used[i]){
                continue;
            }
            used[i] = true;
            accumulate.addLast(nums[i]);
            backTracking(nums,used,accumulate,result);
            accumulate.removeLast();
            used[i] = false;
        }


    }

}